File: python/src/revenant/core/pdf/cms_info.py
# pyright: reportUnknownMemberType=false, reportUnknownVariableType=false, reportUnknownArgumentType=false
"""CMS metadata extraction -- digest info, signer identity, and blob inspection."""

from __future__ import annotations

import hashlib
import logging
from typing import TypedDict

from ...errors import RevenantError
from .asn1 import ASN1_SEQUENCE_TAG, MIN_CMS_SIZE

_logger = logging.getLogger(__name__)

# OID for messageDigest attribute in CMS SignerInfo
_OID_MESSAGE_DIGEST = "1.2.840.113549.1.9.4"

# Map non-standard CMS digest algorithm identifiers to hashlib names.
# CoSign puts sha1WithRSAEncryption in the digestAlgorithm field instead of sha1.
_DIGEST_ALGO_MAP: dict[str, str] = {
    "sha1_rsa": "sha1",
    "sha256_rsa": "sha256",
    "sha384_rsa": "sha384",
    "sha512_rsa": "sha512",
    "1.2.840.113549.1.1.5": "sha1",  # sha1WithRSAEncryption OID
    "1.2.840.113549.1.1.11": "sha256",  # sha256WithRSAEncryption OID
    "1.2.840.113549.1.1.12": "sha384",  # sha384WithRSAEncryption OID
    "1.2.840.113549.1.1.13": "sha512",  # sha512WithRSAEncryption OID
}


def resolve_hash_algo(algo_raw: str) -> str | None:
    """Resolve a CMS digest algorithm identifier to a hashlib-compatible name.

    Args:
        algo_raw: Algorithm name or OID from asn1crypto (.native or .dotted).

    Returns:
        hashlib algorithm name, or None if unrecognized.
    """
    if algo_raw in hashlib.algorithms_available:
        return algo_raw
    return _DIGEST_ALGO_MAP.get(algo_raw)


def extract_digest_info(cms_der: bytes) -> tuple[str, bytes] | None:
    """Extract digest algorithm and messageDigest from a CMS SignerInfo.

    Parses the CMS structure to get:
    1. The digest algorithm from SignerInfo.digestAlgorithm
    2. The messageDigest value from SignerInfo.signedAttrs

    Returns:
        (hashlib_algo_name, digest_bytes) if extraction succeeds, None otherwise.
    """
    try:
        from asn1crypto import cms as asn1_cms

        content_info = asn1_cms.ContentInfo.load(cms_der)
        signed_data = content_info["content"]
        signer_infos = signed_data["signer_infos"]
        if not signer_infos:
            return None

        signer_info = signer_infos[0]

        # Get digest algorithm
        algo_id = signer_info["digest_algorithm"]["algorithm"]
        # Try .native first (e.g. "sha256"), then .dotted OID as fallback
        algo_name = resolve_hash_algo(algo_id.native)
        if algo_name is None:
            algo_name = resolve_hash_algo(algo_id.dotted)
        if algo_name is None:
            _logger.debug("Unrecognized digest algorithm: %s (%s)", algo_id.native, algo_id.dotted)
            return None

        # Get messageDigest from signed attributes
        signed_attrs = signer_info["signed_attrs"]
        if signed_attrs is None:
            return None

        for attr in signed_attrs:
            if attr["type"].dotted == _OID_MESSAGE_DIGEST:
                values = attr["values"]
                if values:
                    return (algo_name, values[0].native)
        return None  # noqa: TRY300 -- not a try success path; this is the "no match found" fallback after the for loop
    except (ValueError, TypeError, KeyError, AttributeError, IndexError):
        _logger.debug("Could not extract digest info from CMS", exc_info=True)
        return None


def extract_signer_info(cms_der: bytes) -> dict[str, str | None] | None:
    """Extract signer certificate info from a CMS blob.

    Returns:
        dict with name, email, organization, dn -- or None on failure.
    """
    try:
        from ..cert_info import extract_cert_info_from_cms

        return extract_cert_info_from_cms(cms_der)
    except (RevenantError, ValueError, TypeError, KeyError, AttributeError):
        _logger.debug("Could not extract signer info from CMS", exc_info=True)
        return None


class CmsInspection(TypedDict):
    """Result of inspecting a CMS/PKCS#7 blob (without original data)."""

    signer: dict[str, str | None] | None
    digest_algorithm: str | None
    cms_size: int
    details: list[str]


def inspect_cms_blob(cms_der: bytes) -> CmsInspection:
    """Inspect a CMS/PKCS#7 blob without verifying against original data.

    Extracts certificate info and digest algorithm. Use this when
    only the .p7s file is available (no original data to verify against).

    Args:
        cms_der: The CMS/PKCS#7 signature (DER-encoded).

    Returns:
        CmsInspection with signer info, digest algorithm, and details.
    """
    details: list[str] = []

    if len(cms_der) < MIN_CMS_SIZE:
        details.append(f"CMS too small ({len(cms_der)} bytes) -- likely corrupt")
        return {
            "signer": None,
            "digest_algorithm": None,
            "cms_size": len(cms_der),
            "details": details,
        }

    if cms_der[0] != ASN1_SEQUENCE_TAG:
        details.append("Not a valid CMS blob (expected ASN.1 SEQUENCE)")
        return {
            "signer": None,
            "digest_algorithm": None,
            "cms_size": len(cms_der),
            "details": details,
        }

    details.append(f"CMS blob: {len(cms_der)} bytes, valid ASN.1 structure")

    signer = extract_signer_info(cms_der)
    if signer:
        if signer.get("name"):
            details.append(f"Signer: {signer['name']}")
        if signer.get("organization"):
            details.append(f"Organization: {signer['organization']}")
        if signer.get("email"):
            details.append(f"Email: {signer['email']}")

    digest_algo = None
    digest_info = extract_digest_info(cms_der)
    if digest_info is not None:
        digest_algo = digest_info[0]
        details.append(f"Digest algorithm: {digest_algo.upper().replace('_', '-')}")

    return {
        "signer": signer,
        "digest_algorithm": digest_algo,
        "cms_size": len(cms_der),
        "details": details,
    }

File: typescript/src/core/pdf/cms-info.ts
// SPDX-License-Identifier: Apache-2.0
/**
 * CMS metadata extraction -- digest info, signer identity, and blob inspection.
 */

import * as asn1js from "asn1js";
import * as pkijs from "pkijs";

import { RevenantError } from "../../errors.js";
import { toArrayBuffer } from "../../utils.js";
import { ASN1_SEQUENCE_TAG, MIN_CMS_SIZE } from "./asn1.js";

/** OID for messageDigest attribute in CMS SignerInfo. */
const OID_MESSAGE_DIGEST = "1.2.840.113549.1.9.4";

/**
 * Map non-standard CMS digest algorithm OIDs to Web Crypto names.
 * CoSign puts sha1WithRSAEncryption in the digestAlgorithm field
 * instead of sha1.
 */
const DIGEST_ALGO_MAP: ReadonlyMap<string, string> = new Map([
  ["1.3.14.3.2.26", "SHA-1"],
  ["2.16.840.1.101.3.4.2.1", "SHA-256"],
  ["2.16.840.1.101.3.4.2.2", "SHA-384"],
  ["2.16.840.1.101.3.4.2.3", "SHA-512"],
  // sha*WithRSAEncryption OIDs (CoSign quirk)
  ["1.2.840.113549.1.1.5", "SHA-1"],
  ["1.2.840.113549.1.1.11", "SHA-256"],
  ["1.2.840.113549.1.1.12", "SHA-384"],
  ["1.2.840.113549.1.1.13", "SHA-512"],
]);

/**
 * Resolve a CMS digest algorithm OID to a Web Crypto name.
 */
export function resolveHashAlgo(algoOid: string): string | null {
  return DIGEST_ALGO_MAP.get(algoOid) ?? null;
}

/**
 * Parse a CMS/PKCS#7 DER blob into a pkijs ContentInfo.
 */
function parseCms(cmsDer: Uint8Array): pkijs.ContentInfo {
  const asn1 = asn1js.fromBER(toArrayBuffer(cmsDer));
  if (asn1.offset === -1) {
    throw new RevenantError("Failed to parse ASN.1 structure from CMS blob");
  }
  return new pkijs.ContentInfo({ schema: asn1.result });
}

/**
 * Extract digest algorithm and messageDigest from a CMS SignerInfo.
 */
export function extractDigestInfo(
  cmsDer: Uint8Array,
): { algorithm: string; digest: Uint8Array } | null {
  try {
    const contentInfo = parseCms(cmsDer);
    const signedData = new pkijs.SignedData({
      schema: contentInfo.content,
    });

    if (signedData.signerInfos.length === 0) return null;

    const signerInfo = signedData.signerInfos[0];
    if (signerInfo === undefined) return null;

    // Get digest algorithm OID
    const algoOid = signerInfo.digestAlgorithm.algorithmId;
    const algoName = resolveHashAlgo(algoOid);
    if (!algoName) return null;

    // Get messageDigest from signed attributes
    const signedAttrs = signerInfo.signedAttrs;
    if (!signedAttrs) return null;

    for (const attr of signedAttrs.attributes) {
      if (attr.type === OID_MESSAGE_DIGEST) {
        if (attr.values.length > 0) {
          const digestValue = attr.values[0];
          if (digestValue !== undefined && digestValue instanceof asn1js.OctetString) {
            const digestBytes = new Uint8Array(digestValue.valueBlock.valueHexView);
            return { algorithm: algoName, digest: digestBytes };
          }
        }
      }
    }

    return null;
  } catch {
    return null;
  }
}

export interface SignerInfo {
  name: string | null;
  email: string | null;
  organization: string | null;
  dn: string | null;
}

/**
 * Extract signer certificate info from a CMS blob.
 * Uses dynamic import to avoid circular dependency with cert-info.
 */
export async function extractSignerInfo(cmsDer: Uint8Array): Promise<SignerInfo | null> {
  try {
    const { extractCertInfoFromCms } = await import("../cert-info.js");
    return extractCertInfoFromCms(cmsDer);
  } catch {
    return null;
  }
}

export interface CmsInspection {
  signer: SignerInfo | null;
  digestAlgorithm: string | null;
  cmsSize: number;
  details: string[];
}

/**
 * Inspect a CMS/PKCS#7 blob without verifying against original data.
 *
 * Extracts certificate info and digest algorithm. Use this when
 * only the .p7s file is available (no original data to verify against).
 */
export async function inspectCmsBlob(cmsDer: Uint8Array): Promise<CmsInspection> {
  const details: string[] = [];

  if (cmsDer.length < MIN_CMS_SIZE) {
    details.push(`CMS too small (${cmsDer.length} bytes) -- likely corrupt`);
    return {
      signer: null,
      digestAlgorithm: null,
      cmsSize: cmsDer.length,
      details,
    };
  }

  if (cmsDer[0] !== ASN1_SEQUENCE_TAG) {
    details.push("Not a valid CMS blob (expected ASN.1 SEQUENCE)");
    return {
      signer: null,
      digestAlgorithm: null,
      cmsSize: cmsDer.length,
      details,
    };
  }

  details.push(`CMS blob: ${cmsDer.length} bytes, valid ASN.1 structure`);

  const signer = await extractSignerInfo(cmsDer);
  if (signer) {
    if (signer.name) details.push(`Signer: ${signer.name}`);
    if (signer.organization) {
      details.push(`Organization: ${signer.organization}`);
    }
    if (signer.email) details.push(`Email: ${signer.email}`);
  }

  let digestAlgorithm: string | null = null;
  const digestInfo = extractDigestInfo(cmsDer);
  if (digestInfo) {
    digestAlgorithm = digestInfo.algorithm;
    details.push(`Digest algorithm: ${digestAlgorithm}`);
  }

  return {
    signer,
    digestAlgorithm,
    cmsSize: cmsDer.length,
    details,
  };
}
